### PR TITLE
Support index cleaner for rollover indices and add integration tests

### LINKED ISSUE #1681: Make es-index-cleaner work with rollover
ES index cleaner does not clean indices created by rollover.

### LINKED ISSUE #1682: Add tests for es-index cleaner and es-rollover
At the moment we don't have any tests for es-index cleaner or rollover it would be great to add a new travis job which test this functionality.

Index cleaner might be harder to test, we can simply create indices manually or report span with old timestamps to create old indices.

For rollover we don't need to report any spans, just to call `rollover` action which creates new indices.

### PR DESCRIPTION
**Pull request by pavolloffay**
Resolves #1681 
Resolves #1682

### CODE REVIEW

**Comment by pavolloffay** [comment]
@objectiser could you please have a look?

File: plugin/storage/es/esCleaner.py [hunk 1]
```diff
@@ -15,7 +15,8 @@ def main():
         print('HOSTNAME ... specifies which Elasticsearch hosts URL to search and delete indices from.')
         print('TIMEOUT ...  number of seconds to wait for master node response.'.format(TIMEOUT))
         print('INDEX_PREFIX ... specifies index prefix.')
-        print('ARCHIVE ... specifies whether to remove archive indices (default false).')
+        print('ARCHIVE ... specifies whether to remove archive indices (only works for rollover) (default false).')
```
Comment (objectiser) [inline plugin/storage/es/esCleaner.py]: Is this a change in behaviour, or did it never work for archive (with daily indices)?

File: plugin/storage/es/esCleaner.py [hunk 2]
```diff
@@ -46,7 +47,10 @@ def main():
     if str2bool(os.getenv("ARCHIVE", 'false')):
         filter_archive_indices(ilo, prefix)
```
Comment (objectiser) [inline plugin/storage/es/esCleaner.py]: nit: based on whether archive only works with rollover, then for consistency might be good to add `_rollover` suffix to function - and may be check for `ROLLOVER` and if not set, then print a warning message?

File: plugin/storage/es/esCleaner.py [hunk 3]
```diff
@@ -60,13 +64,36 @@ def main():
 def filter_main_indices(ilo, prefix):
     ilo.filter_by_regex(kind='prefix', value=prefix + "jaeger")
     empty_list(ilo, "No indices to delete")
+    ilo.filter_by_alias(aliases=[prefix + 'jaeger-span-read'], exclude=True)
```
Comment (objectiser) [inline plugin/storage/es/esCleaner.py]: Why are these included, aren't they related to rollover?

File: plugin/storage/es/esCleaner.py [hunk 4]
```diff
@@ -60,13 +64,36 @@ def main():
...
+    ilo.filter_by_alias(aliases=[prefix + 'jaeger-span-archive-write'], exclude=True)
+    empty_list(ilo, "No indices to delete")
     # This excludes archive index as we use source='name'
     # source `creation_date` would include archive index
     ilo.filter_by_age(source='name', direction='older', timestring='%Y-%m-%d', unit='days', unit_count=int(sys.argv[1]))
 
 
+def filter_main_indices_rollover(ilo, prefix):
+    ilo.filter_by_alias(aliases=[prefix + 'jaeger-*-read'])
+    empty_list(ilo, "No indices to delete")
```
Comment (objectiser) [inline plugin/storage/es/esCleaner.py]: Are these required after each filter, or just at the end? Does it cause problems calling the `filter_by_alias` on an empty list?

File: plugin/storage/es/esCleaner.py [hunk 5]
```diff
@@ -60,13 +64,36 @@ def main():
...
+    ilo.filter_by_alias(aliases=[prefix + 'jaeger-span-write'], exclude=True)
+    empty_list(ilo, "No indices to delete")
+    ilo.filter_by_alias(aliases=[prefix + 'jaeger-service-write'], exclude=True)
+    empty_list(ilo, "No indices to delete")
+    ilo.filter_by_regex(kind='prefix', value=prefix + 'jaeger-span-archive*', exclude=True)
+    empty_list(ilo, "No indices to delete")
+    # This excludes archive index as we use source='name'
+    # source `creation_date` would include archive index
+    # TODO it might be useful to allow filter_by_space
+    ilo.filter_by_age(source='creation_date', direction='older', unit='days', unit_count=int(sys.argv[1]))
```
Comment (objectiser) [inline plugin/storage/es/esCleaner.py]: Just wondering about a particular scenario - if rollover criteria is based on index size (for example) in low activity environment, it might be possible that the same index is in use for more days that the `unit_count`, in which case it would remove the index - even though it is in current use?

File: plugin/storage/integration/es_index_cleaner_test.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+	"fmt"
+	"os/exec"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/olivere/elastic.v5"
+)
+
+// 1. removes daily indices, but does not remove rollover or archive
```
Comment (objectiser) [inline plugin/storage/integration/es_index_cleaner_test.go]: What do these comments relate to? Assume the specific test instances in `TestIndexCleaner` - would be better to move comments to those test instances - and there are three so an additional comment would be required.

File: plugin/storage/es/esCleaner.py (hunk 1 above)
Comment (pavolloffay) [reply plugin/storage/es/esCleaner.py]: It is just a comment change. 

It worked and works for archive - but it is only supported for archive managed by rollover. If there is one archive we do not touch it.

File: plugin/storage/es/esCleaner.py (hunk 3 above)
Comment (pavolloffay) [reply plugin/storage/es/esCleaner.py]: They are excluded, see the last parameter

File: plugin/storage/es/esCleaner.py (hunk 4 above)
Comment (pavolloffay) [reply plugin/storage/es/esCleaner.py]: Yes it's required otherwise it fails on empty cluster

File: plugin/storage/integration/es_index_cleaner_test.go (hunk 6 above)
Comment (pavolloffay) [reply plugin/storage/integration/es_index_cleaner_test.go]: I forgot to remove these

File: plugin/storage/es/esCleaner.py (hunk 3 above)
Comment (objectiser) [reply plugin/storage/es/esCleaner.py]: Sorry poor use of words, I mean why are these lines included/added - as they seem to relate to aliases related to the rollover approach. From looking at the tests, it seems like it is so that existing aliases (from rollover) are not touched - but that would imply the user has gone from a rollover environment back to using daily indices?

File: plugin/storage/es/esCleaner.py (hunk 5 above)
Comment (pavolloffay) [reply plugin/storage/es/esCleaner.py]: I think you are right removing by size does not really make sense.

File: plugin/storage/es/esCleaner.py (hunk 3 above)
Comment (pavolloffay) [reply plugin/storage/es/esCleaner.py]: Yes, I have simplified this in the last commit. Now there is a regex which matches exactly the daily inidices so no need to remove rollover indices afterwards.

File: plugin/storage/es/esCleaner.py [hunk 7]
```diff
@@ -58,17 +62,29 @@ def main():
...
     ilo.filter_by_age(source='name', direction='older', timestring='%Y-%m-%d', unit='days', unit_count=int(sys.argv[1]))
 
 
-def filter_archive_indices(ilo, prefix):
-    # Remove only archive indices when aliases are used
+def filter_main_indices_rollover(ilo, prefix):
+    ilo.filter_by_regex(kind='regex', value=prefix + "jaeger-(span|service)-\d{6}")
+    empty_list(ilo, "No indices to delete")
+    # do not remove active write indices
+    ilo.filter_by_alias(aliases=[prefix + 'jaeger-span-write'], exclude=True)
```
Comment (objectiser) [inline plugin/storage/es/esCleaner.py]: If filtering by the regex now, are these two `aliases` filters required?

File: plugin/storage/es/esCleaner.py [hunk 8]
```diff
@@ -58,17 +62,29 @@ def main():
...
+    empty_list(ilo, "No indices to delete")
+    ilo.filter_by_alias(aliases=[prefix + 'jaeger-service-write'], exclude=True)
+    empty_list(ilo, "No indices to delete")
+    ilo.filter_by_age(source='creation_date', direction='older', unit='days', unit_count=int(sys.argv[1]))
+
+
+def filter_archive_indices_rollover(ilo, prefix):
+    # Remove only rollover archive indices
     # Do not remove active write archive index
     ilo.filter_by_alias(aliases=[prefix + 'jaeger-span-archive-write'], exclude=True)
```
Comment (objectiser) [inline plugin/storage/es/esCleaner.py]: Would it be better to filter by regex here aswell to be consistent with the change in `filter_main_indices_rollover`?

File: plugin/storage/es/esCleaner.py (hunk 7 above)
Comment (pavolloffay) [reply plugin/storage/es/esCleaner.py]: See the comment above. We need to exclude current active write indices.

File: plugin/storage/es/esCleaner.py (hunk 8 above)
Comment (pavolloffay) [reply plugin/storage/es/esCleaner.py]: unfortunatelly this is not `filter_by_regex` method

Comment (objectiser) [reply plugin/storage/es/esCleaner.py]: Not sure what happened - the comment should have been associated with two lines down - so changing the `ilo.filter_by_alias(aliases=[prefix + 'jaeger-span-archive-read'])` to be `ilo.filter_by_regex(kind='regex', value=prefix + "jaeger-(span|service)-\d{6}")` ?

Comment (pavolloffay) [reply plugin/storage/es/esCleaner.py]: Sorry,  I do not understand what you mean.

1. we use regex to find indices which should be removed
2. but we also need to exclude indices which are associated with write alias - to do that we use filter for an alias.

Comment (objectiser) [reply plugin/storage/es/esCleaner.py]: Just mean that `filter_main_indices` and `filter_main_indices_rollover` are both using `filter_by_regex` to identify the indices to be removed - whereas `filter_archive_indices_rollover` uses `ilo.filter_by_alias(aliases=[prefix + 'jaeger-span-archive-read'])` - just suggesting that it would be better if `filter_archive_indices_rollover` also uses `filter_by_regex` for consistency.

Comment (pavolloffay) [reply plugin/storage/es/esCleaner.py]: Yes, that sounds better https://github.com/jaegertracing/jaeger/pull/1693
